### ISSUE TITLE
opensslconf.h and opensslv.h include each other

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#7874]
The subject line says it all.  These two files should be untangled somehow.  One option, probably the simplest (I spent a bit of time trying to do it and failed) is to merge opensslv.h into opensslconf.h.in and make the first one just be a one-liner that includes the second one.

**Comment by levitte** [comment]
Hmmm, how about something less disruptive?  We could move all the API compat stuff to `opensslv.h`, that way it won't need to include `opensslconf.h`

**Comment by richsalz** [comment]
I couldn't make that work, but you're smarter than I, so go for it :)

On the other rand, right now each file includes the other, so the simpler change won't break user-code that depends on that.

**Comment by levitte** [comment]
I've a question: have you observed an actual issue, such as compilation going berzerk, dependency errors or something like that?  I.e., pragmatically speaking, is there anything more concrete than your own sense of what's right and beautiful?

I'm asking 'cause this can be observed:

opensslv.h:
``` C
#ifndef HEADER_OPENSSLV_H
# define HEADER_OPENSSLV_H

...

# include <openssl/opensslconf.h>

...

#endif                          /* HEADER_OPENSSLV_H */
```

opensslconf.h:
``` C
#include <openssl/opensslv.h>

#ifndef HEADER_OPENSSLCONF_H
# define HEADER_OPENSSLCONF_H

...

#endif                          /* HEADER_OPENSSLCONF_H */
```

In other words, at least one of the includes is done under the protection of guards, and that prevents inclusion loops.  Can you see a scenario where that prevension is circumvented?

**Comment by levitte** [comment]
Scenario 1:
- someone includes opensslv.h
- opensslv.h defines `HEADER_OPENSSLV_H`
- opensslv.h includes opensslconf.h
    - opensslconf.h includes opensslv.h, whose contents are ignored because of `HEADER_OPENSSLV_H`
    - opensslconf.h defines HEADER_OPENSSLCONF_H
    - opensslconf.h finishes
- opensslv.h finishes

Scenario 2:
- someone includes opensslconf.h
- opensslconf.h includes opensslv.h
    - opensslv.h defines `HEADER_OPENSSLV_H`
    - opensslv.h includes opensslconf.h
        - opensslconf.h includes opensslv.h, whose contents are ignored because of `HEADER_OPENSSLV_H`
        - opensslconf.h defines HEADER_OPENSSLCONF_H
        - opensslconf.h finishes
    - opensslv.h finishes
- opensslconf.h is skipped over because the earlier inclusion by opensslv.h defined HEADER_OPENSSLCONF_H

So yeah, there are a couple of levels of inclusion, but nothing that makes it loop endlessly.

Worth fixing, or shall we invoke the "it works, don't fix it!" mantra?

**Comment by richsalz** [comment]
> Worth fixing, or shall we invoke the "it works, don't fix it!" mantra?

Feel free to close the issue if you don't want to address it.

I found it trying to add a guard to the other header file.  It "stuck out" that there was no guard there, and then I fell down the rabbit-hole trying to fix it.

**Comment by levitte** [comment]
> I found it trying to add a guard to the other header file

Uhmmmm, which "other header file" are you referring to?

**Comment by richsalz** [comment]
Oh never mind.  I was hopelessly lost.  Just close this darn thing if you don't care.

**Comment by levitte** [comment]
Yeah, I think I'll do so.  "It works, don't fix it!" is a good mantra at times.

Of course, if we do find a concrete case where it doesn't work, it will be time to revisit the issue.

**Comment by mspncp** [comment]
> Hmmm, how about something less disruptive? We could move all the API compat stuff to `opensslv.h`, that way it won't need to include `opensslconf.h`

@richsalz was right, it i causing problems.  And @levitte's first reaction would have bee the appropriate solution.

Fix is now in #9623.
